File: sdk/ruby/lib/shojiku/engine.rb
# frozen_string_literal: true

require "fiddle"

module Shojiku
  # Everything copied out of one result handle, before that handle is freed.
  #
  # A snapshot rather than a wrapper, and that is the ownership rule of this
  # binding in one word: no Ruby object ever holds a pointer into engine
  # memory. The accessors LEND — their pointers die with the handle — so the
  # bytes are copied while the handle is alive and the handle is freed on the
  # way out, on every path.
  Snapshot = Data.define(:status, :success, :pdf, :json, :diagnostics, :error)

  # The declared C surface, and the one place a call crosses into it.
  #
  # Every function is declared with explicit argument and return types.
  # Fiddle's default return type is a C `int`, which truncates every pointer
  # this library hands back — a segfault that looks like a memory bug and is
  # really a missing declaration.
  class Engine
    VOIDP = Fiddle::TYPE_VOIDP
    SIZE_T = Fiddle::TYPE_SIZE_T
    INT = Fiddle::TYPE_INT
    VOID = Fiddle::TYPE_VOID

    # Unpack directives that match the C types EXACTLY, rather than Ruby's
    # native-width shorthands. `l!` is a native `long`, which is 8 bytes where
    # `int32_t` is 4 — and `unpack1` on a buffer shorter than its directive
    # returns nil rather than raising, so every flag would silently read as
    # false. `l` is int32 and these two are picked from the real `size_t`
    # width, which differs from `unsigned long` on Windows.
    INT32 = "l"
    SIZE = Fiddle::SIZEOF_SIZE_T == 8 ? "Q" : "L"

    # Every data argument crosses as a (pointer, length) PAIR — nothing is
    # NUL-terminated, because PDF bytes contain NUL — and every call ends with
    # the out-slot the result handle is written to. Spelling the signatures as
    # a table rather than one call each keeps them impossible to disagree
    # with the header: an operation's arity is the count of its arguments.
    PAIR = [VOIDP, SIZE_T].freeze
    OUT = [VOIDP].freeze

    # Only the lifecycle the SDK contract defines is bound: engine info,
    # render, sign, verify. `validate` and `preview` are the authoring
    # surface's, not an artifact lifecycle's — the Designer reaches them
    # through the WASM bindings, and binding them here would be surface with
    # no contract behind it.
    SIGNATURES = {
      shojiku_engine_info: OUT,
      shojiku_render: PAIR + OUT,
      shojiku_sign: (PAIR * 4) + OUT,
      shojiku_sign_prepare: (PAIR * 3) + OUT,
      shojiku_sign_complete: (PAIR * 4) + OUT,
      shojiku_verify: (PAIR * 2) + OUT
    }.freeze

    def initialize(library)
      @library = library
      @calls = SIGNATURES.to_h { |name, args| [name, library.function(name, args, INT)] }
      declare_accessors(library)
    end

    def engine_info
      invoke { |out| call(:shojiku_engine_info, out) }
    end

    def render(request)
      invoke { |out| call(:shojiku_render, request, request.bytesize, out) }
    end

    def sign(pdf:, key:, certificate:, passphrase: nil)
      invoke do |out|
        call(:shojiku_sign, pdf, pdf.bytesize, key, key.bytesize,
             certificate, certificate.bytesize,
             passphrase, passphrase ? passphrase.bytesize : 0, out)
      end
    end

    # The two halves of signing with a key held elsewhere. Both take the same
    # document, certificate and algorithm — the pair is stateless, so the
    # second call re-derives what the first prepared.
    def sign_prepare(pdf:, certificate:, algorithm:)
      invoke do |out|
        call(:shojiku_sign_prepare, pdf, pdf.bytesize, certificate, certificate.bytesize,
             algorithm, algorithm.bytesize, out)
      end
    end

    def sign_complete(pdf:, certificate:, algorithm:, signature:)
      invoke do |out|
        call(:shojiku_sign_complete, pdf, pdf.bytesize, certificate, certificate.bytesize,
             algorithm, algorithm.bytesize, signature, signature.bytesize, out)
      end
    end

    def verify(pdf:, anchors:)
      invoke { |out| call(:shojiku_verify, pdf, pdf.bytesize, anchors, anchors.bytesize, out) }
    end

    private

    # The one place a lifecycle call crosses.
    def call(name, *)
      @calls.fetch(name).call(*)
    end

    def declare_accessors(library)
      buffers = %i[shojiku_result_pdf shojiku_result_json shojiku_result_diagnostics_json
                   shojiku_result_error_json]
      @buffers = buffers.to_h { |name| [name, library.function(name, [VOIDP] * 3, INT)] }
      @success = library.function(:shojiku_result_success, [VOIDP, VOIDP], INT)
      @free = library.function(:shojiku_result_free, [VOIDP], VOID)
    end

    # Runs one operation and copies its result out.
    #
    # The `ensure` is the ownership contract: exactly one handle crosses and
    # exactly one free pairs with it, whatever happens in between.
    def invoke
      out = zeroed(Fiddle::SIZEOF_VOIDP)
      status = yield(out)
      handle = out.ptr
      begin
        snapshot(status, handle)
      ensure
        @free.call(handle)
      end
    end

    def snapshot(status, handle)
      Snapshot.new(
        status: status,
        success: succeeded?(handle),
        pdf: buffer(handle, :shojiku_result_pdf),
        json: text(handle, :shojiku_result_json),
        diagnostics: text(handle, :shojiku_result_diagnostics_json),
        error: text(handle, :shojiku_result_error_json)
      )
    end

    def succeeded?(handle)
      slot = zeroed(Fiddle::SIZEOF_INT)
      @success.call(handle, slot)
      slot[0, Fiddle::SIZEOF_INT].unpack1(INT32) == 1
    end

    # Copies what an accessor lent. `to_str` copies, which is the whole point:
    # the pointer it copies from stops being valid the moment the handle is
    # freed, a few lines later.
    def buffer(handle, name)
      pointer = zeroed(Fiddle::SIZEOF_VOIDP)
      length = zeroed(Fiddle::SIZEOF_SIZE_T)
      @buffers.fetch(name).call(handle, pointer, length)
      size = length[0, Fiddle::SIZEOF_SIZE_T].unpack1(SIZE)
      size.zero? ? (+"").b : pointer.ptr.to_str(size)
    end

    # The same, for a buffer the surface guarantees is UTF-8. The encoding is
    # forced rather than inherited: a platform default would differ on
    # Windows, which is a first-class target here.
    def text(handle, name)
      buffer(handle, name).force_encoding(Encoding::UTF_8)
    end

    # Out-parameters start at zero, so a slot the library never wrote reads
    # as "absent" rather than as whatever was in that memory.
    def zeroed(size)
      slot = Fiddle::Pointer.malloc(size, Fiddle::RUBY_FREE)
      slot[0, size] = "\x00" * size
      slot
    end
  end
end

File: sdk/java/src/main/java/jp/kengos/shojiku/Engine.java
package jp.kengos.shojiku;

import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import java.nio.charset.StandardCharsets;

/**
 * The one place a call crosses into the engine, and the copy-then-free discipline around it.
 *
 * <p>Only the lifecycle the SDK contract defines is bound: engine info, render, sign, verify.
 * {@code validate} and {@code preview} are the authoring surface's, not an artifact lifecycle's —
 * the Designer reaches them through the WASM bindings, and binding them here would be surface with
 * no contract behind it.
 */
final class Engine {

  private static final byte[] NOTHING = new byte[0];

  private final EngineLibrary library;
  private final ShojikuLibrary bound;

  Engine(EngineLibrary library) {
    this.library = library;
    this.bound = library.bound();
  }

  Snapshot engineInfo() {
    PointerByReference out = new PointerByReference();
    return read(bound.shojiku_engine_info(out), out);
  }

  Snapshot render(byte[] request) {
    PointerByReference out = new PointerByReference();
    int status = bound.shojiku_render(request, size(request), out);
    return read(status, out);
  }

  Snapshot sign(byte[] pdf, byte[] key, byte[] certificate, byte[] passphrase) {
    PointerByReference out = new PointerByReference();
    // A null passphrase is what the header wants for an unencrypted key; an
    // empty array would be a zero-length passphrase, which is a different claim.
    byte[] pass = passphrase == null || passphrase.length == 0 ? null : passphrase;
    int status =
        bound.shojiku_sign(
            pdf,
            size(pdf),
            key,
            size(key),
            certificate,
            size(certificate),
            pass,
            new SizeT(pass == null ? 0 : pass.length),
            out);
    return read(status, out);
  }

  Snapshot verify(byte[] pdf, byte[] anchors) {
    PointerByReference out = new PointerByReference();
    int status = bound.shojiku_verify(pdf, size(pdf), anchors, size(anchors), out);
    return read(status, out);
  }

  /**
   * Copy one result out, then free it.
   *
   * <p>The {@code finally} is the ownership contract: exactly one handle crosses and exactly one
   * free pairs with it, whatever happens in between. A blank out-slot — which is what the header
   * leaves on a rejected call — is not dereferenced at all; the status already says what happened,
   * and freeing NULL is a documented no-op.
   */
  Snapshot read(int status, PointerByReference out) {
    Pointer handle = out.getValue();
    if (handle == null) {
      return new Snapshot(status, false, NOTHING, "", "", "");
    }
    try {
      return new Snapshot(
          status,
          succeeded(handle),
          buffer(handle, Accessor.PDF),
          text(handle, Accessor.JSON),
          text(handle, Accessor.DIAGNOSTICS),
          text(handle, Accessor.ERROR));
    } finally {
      bound.shojiku_result_free(handle);
    }
  }

  private boolean succeeded(Pointer handle) {
    IntByReference slot = new IntByReference();
    bound.shojiku_result_success(handle, slot);
    return slot.getValue() == 1;
  }

  /**
   * Copy what an accessor lent.
   *
   * <p>A copy, which is the whole point: the pointer it copies from stops being valid the moment
   * the handle is freed, a few lines later. The length is read FIRST, so an empty buffer never
   * dereferences the pointer beside it.
   */
  private byte[] buffer(Pointer handle, Accessor accessor) {
    PointerByReference pointer = new PointerByReference();
    SizeTByReference length = new SizeTByReference();
    accessor.read(bound, handle, pointer, length);
    long size = length.getValue();
    if (size == 0) {
      return NOTHING;
    }
    return pointer.getValue().getByteArray(0, (int) size);
  }

  /**
   * The same, for a buffer the surface guarantees is UTF-8.
   *
   * <p>Decoded explicitly rather than by whatever the platform would pick: Windows is a first-class
   * target here and its default differs.
   */
  private String text(Pointer handle, Accessor accessor) {
    return new String(buffer(handle, accessor), StandardCharsets.UTF_8);
  }

  private static SizeT size(byte[] bytes) {
    return new SizeT(bytes.length);
  }

  /** The four buffer accessors, which all share one signature. */
  private enum Accessor {
    PDF {
      @Override
      void read(
          ShojikuLibrary bound, Pointer handle, PointerByReference out, SizeTByReference length) {
        bound.shojiku_result_pdf(handle, out, length);
      }
    },
    JSON {
      @Override
      void read(
          ShojikuLibrary bound, Pointer handle, PointerByReference out, SizeTByReference length) {
        bound.shojiku_result_json(handle, out, length);
      }
    },
    DIAGNOSTICS {
      @Override
      void read(
          ShojikuLibrary bound, Pointer handle, PointerByReference out, SizeTByReference length) {
        bound.shojiku_result_diagnostics_json(handle, out, length);
      }
    },
    ERROR {
      @Override
      void read(
          ShojikuLibrary bound, Pointer handle, PointerByReference out, SizeTByReference length) {
        bound.shojiku_result_error_json(handle, out, length);
      }
    };

    abstract void read(
        ShojikuLibrary bound, Pointer handle, PointerByReference out, SizeTByReference length);
  }
}
